### PR TITLE
all: rename NetState::get_state() to ::retrieve()

### PR DESCRIPTION
**Pull request by acardace**
Use a retrieve() function with an impl block with is more rust idiomatic.

Signed-off-by: Antonio Cardace <acardace@redhat.com>

### CODE REVIEW

**Comment by cathay4t** [comment]
I must read the rust document wrong. From my view, the `NetState::new()` is creating a object in memory not retrieving network state from kernel.

And `NetState::new()` does not handle this future design:

```
state = NetState::new(); // Empty state
state.ifaces["eth1"] = EthernetIface::new("eth1");
state.ifaces["eth1"].mtu = 1280
state.apply()
```

If you think your proposal is rusty way, please provide rust document indication so.

**Comment by cathay4t** [comment]
According to [Rust API Guideline][1], the `Class::new()`:

```
The name new should generally be used for the primary method of instantiating a type
```

For this PR, I would suggest `NetState::retrieve() -> Result<NetState, NisporError>`.


[1]: https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor

**Comment by acardace** [comment]
> I must read the rust document wrong. From my view, the `NetState::new()` is creating a object in memory not retrieving network state from kernel.
> 
> And `NetState::new()` does not handle this future design:
> 
> ```
> state = NetState::new(); // Empty state
> state.ifaces["eth1"] = EthernetIface::new("eth1");
> state.ifaces["eth1"].mtu = 1280
> state.apply()
> ```
> 
> If you think your proposal is rusty way, please provide rust document indication so.

Well it was just an indication of how I usually see things done in rust, I was not aware of this future design, but anyway I have to say it looks good to me.

Maybe better with
```
state.add_link(EthernetIface::new("eth1"));
``` 
But I am not that experienced with rust.

**Comment by acardace** [comment]
> According to [Rust API Guideline](https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor), the `Class::new()`:
> 
> ```
> The name new should generally be used for the primary method of instantiating a type
> ```
> 
> For this PR, I would suggest `NetState::retrieve() -> Result<NetState, NisporError>`.

`retrieve()` looks good as well but `new()` in this PR fits exactly the case of instantiating a new type, doesn't it?

**Comment by cathay4t** [comment]
> > According to [Rust API Guideline](https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor), the `Class::new()`:
> > ```
> > The name new should generally be used for the primary method of instantiating a type
> > ```
> > 
> > 
> > For this PR, I would suggest `NetState::retrieve() -> Result<NetState, NisporError>`.
> 
> `retrieve()` looks good as well but `new()` in this PR fits exactly the case of instantiating a new type, doesn't it?

Normally, the `new()` function does not generate failures but a new instance which does not fit into this use case as we
are expecting `NisporError` to be raised when terrible failure found.

**Comment by acardace** [comment]
> > > According to [Rust API Guideline](https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor), the `Class::new()`:
> > > ```
> > > The name new should generally be used for the primary method of instantiating a type
> > > ```
> > > 
> > > 
> > > For this PR, I would suggest `NetState::retrieve() -> Result<NetState, NisporError>`.
> > 
> > 
> > `retrieve()` looks good as well but `new()` in this PR fits exactly the case of instantiating a new type, doesn't it?
> 
> Normally, the `new()` function does not generate failures but a new instance which does not fit into this use case as we
> are expecting `NisporError` to be raised when terrible failure found.

Mhmm, so maybe Netstate should have a `new()` function to create the object and a `retrieve()` to be filled with infos from netlink, what do you think? Otherwise you can just close this.

**Comment by cathay4t** [comment]
> > > > According to [Rust API Guideline](https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor), the `Class::new()`:
> > > > ```
> > > > The name new should generally be used for the primary method of instantiating a type
> > > > ```
> > > > 
> > > > 
> > > > For this PR, I would suggest `NetState::retrieve() -> Result<NetState, NisporError>`.
> > > 
> > > 
> > > `retrieve()` looks good as well but `new()` in this PR fits exactly the case of instantiating a new type, doesn't it?
> > 
> > 
> > Normally, the `new()` function does not generate failures but a new instance which does not fit into this use case as we
> > are expecting `NisporError` to be raised when terrible failure found.
> 
> Mhmm, so maybe Netstate should have a `new()` function to create the object and a `retrieve()` to be filled with infos from netlink, what do you think? Otherwise you can just close this.

We don't allow user to create `NetState` yet. So there is no need of `NetState::new()`. But I am OK to use `NetState::retrieve()`
for state retrieving.

You may continue work on this PR or create a new one(please also fix test and python binding).

**Comment by acardace** [comment]
> > > > > According to [Rust API Guideline](https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor), the `Class::new()`:
> > > > > ```
> > > > > The name new should generally be used for the primary method of instantiating a type
> > > > > ```
> > > > > 
> > > > > 
> > > > > For this PR, I would suggest `NetState::retrieve() -> Result<NetState, NisporError>`.
> > > > 
> > > > 
> > > > `retrieve()` looks good as well but `new()` in this PR fits exactly the case of instantiating a new type, doesn't it?
> > > 
> > > 
> > > Normally, the `new()` function does not generate failures but a new instance which does not fit into this use case as we
> > > are expecting `NisporError` to be raised when terrible failure found.
> > 
> > 
> > Mhmm, so maybe Netstate should have a `new()` function to create the object and a `retrieve()` to be filled with infos from netlink, what do you think? Otherwise you can just close this.
> 
> We don't allow user to create `NetState` yet. So there is no need of `NetState::new()`. But I am OK to use `NetState::retrieve()`
> for state retrieving.
> 
> You may continue work on this PR or create a new one(please also fix test and python binding).

Ok I updated the PR, I should have addressed everything!

File: src/clib/lib.rs [hunk 1]
```diff
@@ -36,7 +36,7 @@ pub extern "C" fn nispor_state_get(
 }
 
 #[no_mangle]
-pub extern "C" fn nispor_state_free(state: *mut c_char) {
+pub extern "C" fn nispor_retrieve_free(state: *mut c_char) {
```
Comment (cathay4t) [inline src/clib/lib.rs]: How about `nispor_net_state_free()`?

File: src/clib/lib.rs [hunk 2]
```diff
@@ -4,7 +4,7 @@ use serde_json;
 use std::ffi::CString;
 
 #[no_mangle]
-pub extern "C" fn nispor_state_get(
+pub extern "C" fn nispor_retrieve(
```
Comment (cathay4t) [inline src/clib/lib.rs]: How about keep the old name or use `nispor_net_state_retrieve` which better align with rust API.

File: src/python/nispor/state.py [hunk 3]
```diff
@@ -37,5 +37,5 @@ def routes(self):
         return self._routes
 
 
-def get_state():
-    return NisporState(json.loads(get_state_json()))
+def retrieve():
```
Comment (cathay4t) [inline src/python/nispor/state.py]: How about we create a python struct `NetState` to be 1:1 matching with rust one?

**Review by cathay4t — CHANGES_REQUESTED** [review]

File: src/python/nispor/state.py (hunk 3 above)
Comment (cathay4t) [reply src/python/nispor/state.py]: How about place it as static method as `NisporNetState.retrieve()`?

File: src/python/nispor/clib_wrapper.py [hunk 4]
```diff
@@ -39,17 +39,17 @@ def __init__(self, kind, msg):
         super().__init__(f"{kind}: {msg}")
 
 
-def get_state_json():
+def retrieve_json():
```
Comment (cathay4t) [inline src/python/nispor/clib_wrapper.py]: How about `retrieve_net_state_json()`?

**Review by cathay4t — CHANGES_REQUESTED** [review]

File: src/python/nispor/clib_wrapper.py (hunk 4 above)
Comment (acardace) [reply src/python/nispor/clib_wrapper.py]: Done

File: src/python/nispor/state.py (hunk 3 above)
Comment (cathay4t) [reply src/python/nispor/state.py]: The function `retrieve()` is still as module level rather than at `NisporNetState` class level.

Comment (acardace) [reply src/python/nispor/state.py]: Sorry forgot, now it should be ok.

**Review by cathay4t — APPROVED** [review]

**Comment by cathay4t** [comment]
Merged. Thanks.

**Pull request merged**